pkg/apis/hive/v1alpha1/syncset_types.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// Sync SyncSetApplyMode indicates that objects will be deleted
+	// if detected missing from defined Resources in the SyncSet.
+	Sync SyncSetApplyMode = 0
+
+	// Upsert SyncSetApplyMode indicates that objects will be updated
+	// or inserted (created).
+	Upsert SyncSetApplyMode = 1
+)
+
+// SyncSetApplyMode is an enum representing the mode with which to sync objects.
+type SyncSetApplyMode int
+
+// SyncObjectPatch represents a patch to be applied to a specific object.
+type SyncObjectPatch struct {
+	GroupVersionKind schema.GroupVersionKind `json:"groupVersionKind"`
+	Name             string                  `json:"name"`
+
+	// Namespace is the Namespace in which the object to patch exists.
+	// Defaults to the SyncSet's Namespace.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Patch is the patch to apply.
+	Patch []byte `json:"patch"`
+
+	// PatchType indicates the PatchType as JSONPatchType (default),
+	// MergePatchType, or StrategicMergePatchType.
+	// +optional
+	PatchType types.PatchType `json:"patchType,omitempty"`
+}
+
+// SyncDescriptor describes objects that have been created or patches that
+// have been applied using the unique md5 sum of the object or patch.
+type SyncDescriptor struct {
+	GroupVersionKind schema.GroupVersionKind `json:"groupVersionKind"`
+	Name             string                  `json:"name"`
+	Namespace        string                  `json:"namespace"`
+	Hash             []byte                  `json:"hash"`
```

Comment:
Will the hash be helpful now? I'd be ok with skipping it if we don't think we need it for anything just yet. I don't think we can use it to determine if something needs to be done or not because we need to check remote state, and the hash there won't match due to many things like creationTimestamp.